service/history/historyEngine.go

```diff
@@ -691,6 +691,11 @@ func (e *historyEngineImpl) startWorkflowHelper(
 				startRequest,
 				nil,
 			)
+			// By the time we try to terminate the workflow, it was already terminated
+			// So continue as if we didn't need to terminate it in the first place
+			if _, ok := err.(*types.WorkflowExecutionAlreadyCompletedError); !ok {
+				return resp, err
```

Comment:
@yux0 Can you double check the logic here? Looks good to me, but if possible please double check.

Reply:
Technically, this is correct. However, we acquire locking before shouldTerminateAndStart. So the workflow state should not be changed. Maybe we can add a log meesage if the AlreadyCompletedError encounters.

Reply:
Added a warning log, let me know if this looks good to you